### ISSUE TITLE
`MinimumGridSpacingCompute` is slow and does not appear to be memoized properly.

### ISSUE THREAD

**Issue by nikwit** [sxs-collaboration/spectre#3322]
# Bug reports:

When profiling my curved scalar wave executable I find that over 20% of execution time is spent in `minimum_grid_spacing`. 
I can reproduce this behavior with the `EvolveScalarWavePlaneWave3D` from current develop. 

Firstly, why is this function so incredible slow? I faintly remember this coming up a few months ago but don't remember the details. Is it possible that the `goto` statement in there prevents compiler optimizations?

Secondly, e.g. for `EvolveScalarWavePlaneWave3D` the inertial coordinates should be time-independent, right? Why does the databox not cache the result properly given that the input arguments (the mesh and the coordinates) should in theory not be changing?

### Environment:

Add as an attachment `$SPECTRE_BUILD_DIR/LibraryVersions.txt` or
add its contents here.

# Feature request:

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Desired feature:

- [ ] Detail 1
- [ ] Detail 2
- [ ] Detail 3

**Comment by fmahebert** [comment]
When I previously looked into this a bit, the most promising thing I found was that the `operator==` for `std::array` was slower than explicitly checking by hand `if constexpr (Dim == 1) { return lhs[0] == rhs[0] } else if constexpr ...` and so on.

But I never got a satisfactory understanding of why the function remained so expensive. There is quite a bit of looping done, comparing lots of pairs of points... could speed that up by only comparing points near the boundary (where the points cluster), but that's making an assumption that _could_ fail. And to allow for time-dependent maps, you need to redo the work each time step. But even considering all this, it's still surprisingly expensive...

**Comment by wthrowe** [comment]
> When I previously looked into this a bit, the most promising thing I found was that the `operator==` for `std::array` was slower than explicitly checking by hand `if constexpr (Dim == 1) { return lhs[0] == rhs[0] } else if constexpr ...` and so on.

Interesting.  libstdc++ implements this operation using `std::equal`.  I don't see any reason that should be slow (maybe the compiler can't figure out how to unroll the loop?), but writing a comparison that has no loops is straightforward: (untested)
```c++
template <typename T, size_t N, size_t... Is>
bool array_equal_impl(const std::array<T, N>& a, const std::array<T, N>& b,
                      std::index_sequence<Is...> /*meta*/) noexcept {
  return (a[Is] == b[Is] and ...);
}
template <typename T, size_t N>
bool array_equal(const std::array<T, N>& a, const std::array<T, N>& b) noexcept {
  return array_equal_impl(a, b, std::make_index_sequence<N>{});
}
```

> But I never got a satisfactory understanding of why the function remained so expensive. There is quite a bit of looping done, comparing lots of pairs of points... could speed that up by only comparing points near the boundary (where the points cluster), but that's making an assumption that _could_ fail. And to allow for time-dependent maps, you need to redo the work each time step. But even considering all this, it's still surprisingly expensive...

Could also be inefficiencies form using IndexIterator.  I don't know how bad that is, but if nothing else it will prevent any loops from being unrolled.

It might also be acceptable for this to return an approximate value, even going as far as assuming the smallest is on the boundary and orthogonal to it.  That's probably right as long as the grid is skewed by less than 30 degrees.  (This might be violated by some domains, but hopefully not by much.)  This value is mainly used for CFL adjustments to the time-step size, which are approximate anyway.

**Comment by moxcodes** [comment]
In a previous conversation with @nilsdeppe and @fmahebert we also had the thought that it would probably be a decent approximation to go so far as to assume that the smallest grid spacing would be on corners of elements, based on a similar reasoning of 'not too distorted' maps. Then, we'd only ever check dim * 2^(dim) point distances.

**Comment by nilsdeppe** [comment]
Yes, I think whenever that assumption fails it means the map is varying so rapidly within the element that we can't resolve it with the basis functions anyway :) 

I'd advocate for adding an `ApproximateMinimumGridSpacingCompute` that can be swapped in for the `MinimumGridSpacingCompute` (inherits from `MinimumGridSpacing`) that has both the optimizations.

**Comment by nikwit** [comment]
Any idea about the memoization?

**Comment by nilsdeppe** [comment]
That's complicated because you can specify time-dependence at runtime. Generally the safest way to do that would be to use a `std::optional<bool>`, get the time-dependent map, and then decide if you've already computed the correct number, but you also need to be aware that something like AMR can change the grid in the future... I would defer this and first do the other optimization since that might be enough

**Comment by moxcodes** [comment]
I think that if we are using time-dependent maps, then it's harder to re-use the result. I would have thought that the current use of this tag as a compute tag should 'implicitly' memoize whenever the domain was static -- in the sense that it wouldn't recompute the grid spacing unless one of the inputs changed, but perhaps there's something going wrong where a dependency is getting mutated when it doesn't need to be?

**Comment by fmahebert** [comment]
I like the idea of adding an optimized, approximate version of this compute item as a new `ApproximateMinimumGridSpacingCompute`.

Re @wthrowe's comment about assuming the smallest spacing is orthogonal to the boundary (for small skew): I agree in principle, but some of our maps do have larger skews. For instance, the cubed sphere map can have 45 degrees of skew in two different directions, and if I did the geometry right this means the minimum grid spacing is to an "opposite" corner rather than on the boundaries.

**Comment by moxcodes** [comment]
I guess we could do the corners-of-elements approximation, but check not only the adjacent points, but the 'nearest' (2^dim - 1) points that includes corner-adjacency. That would be (2^(2 dim) - 2^dim) comparisons, which still scales far better than the current implementation.
~~I actually think the current implementation only checks adjacent points along each dimension, so also fails in the case of 45-degree skew.~~ edit: nevermind, I'm mistaken

**Comment by kidder** [comment]
Note `MinimumGridSpacingCompute` depends on  `::Tags::time` 

via `Coordinates<Frame::Inertial>` which is provided by  `::domain::Tags::InertialFromGridCoordinatesCompute<Dim>` which depends upon `CoordinatesMeshVelocityAndJacobians<Dim>` which is provided by `::domain::Tags::CoordinatesMeshVelocityAndJacobiansCompute<
          ::domain::CoordinateMaps::Tags::CoordinateMap<Dim, Frame::Grid,
                                                        Frame::Inertial>>` which depends on `::Tags::Time`

**Comment by wthrowe** [comment]
Checking distances from the corner points will fail for the "inverse cubed sphere" (sphere on the inside, cube on the outside), but so will every other idea we've come up with that doesn't do a full check of all adjacencies of all border points.

**Comment by wthrowe** [comment]
We have the not-yet-widely-used `enable_time_dependent_maps` metavariables parameter.  We could probably use that to change the compute tag dependencies and avoid recomputing in the case of a static domain.

**Comment by nilsdeppe** [comment]
Yea, that would work if we want to have an executable not be able to use time dependences at all, but I'm not sure how far that's been pushed in terms of support in all the different maps and domain creators
